lib/index.d.ts

```diff
@@ -0,0 +1,99 @@
+type i53 = number;
+type i24 = number;
+
+type Ty = Array<i53>;
+
+type Column = Array<unknown>;
+
+type Archetype = {
+	id: number,
+	edges: {
+		[key: i53]: {
+			add: Archetype,
+			remove: Archetype,
+		},
+	},
+	types: Ty,
+	type: string | number,
+	entities: Array<number>,
+	columns: Array<Column>,
+	records: { [key: number]: number },
+}
+
+type ArchetypeMap = {
+	cache: Array<ArchetypeRecord>,
+	first: ArchetypeMap,
+	second: ArchetypeMap,
+	parent: ArchetypeMap,
+	size: number,
+}
+
+type ArchetypeRecord = number;
+
+export type Entity<T = unknown> = number & { __nominal_type_dont_use: T }
+
+type EntityIndex = {
+    dense: {
+        [key: i24]: i53
+    };
+    sparse: {
+        [key: i53]: Record
+    }
+}
+
+type Record = {
+	archetype: Archetype,
+	row: number,
+	dense: i24,
+	componentRecord: ArchetypeMap,
+}
+
+type ExtractFromLuaTuple<T> = T extends LuaTuple<infer U> ? U : never;
+
+type QueryShim<T extends any[]> = {
+    without: (...args: ExtractFromLuaTuple<T>) => QueryShim<T>;
+} & IterableFunction<T>;
+
+interface World {
+    entity(): Entity;
+    component<T = unknown>(): Entity<T>;
+
+    target(id: Entity, relation: Entity): Entity | undefined;
+    delete(id: Entity): void;
+
+    add<T>(id: Entity, component: Entity<T>): void;
+    set<T>(id: Entity, component: Entity<T>, data: T): void;
+    remove(id: Entity, component: Entity): void;
+
+    get<A>(id: number, component: Entity<A>): A; // Manually typed out since there is a hard limit.
+    get<A, B>(id: number, component: Entity<A>, component2: Entity<B>): LuaTuple<[A, B]>;
+    get<A, B, C>(id: number, component: Entity<A>, component2: Entity<B>, component3: Entity<C>): LuaTuple<[A, B, C]>;
+    get<A, B, C, D>(id: number, component: Entity<A>, component2: Entity<B>, component3: Entity<C>, component4: Entity<D>): LuaTuple<[A, B, C, D]>;
+    
+    query<T extends unknown[]>(...entities: T): QueryShim<LuaTuple<[...T]>>

```

Comment:
Similar thing here that `T` should extend to `Entity[]` and the function overall should be `query<T extends Entity[]>(...components: T): Query<InferComponents<T>>`. We call them `components` because the entities are being used as components in this case which is fine.

The suggestion for how `InferComponents` should look like would be something like this:

```ts
export type Type<T> = T extends Entity<infer A> ? A : never;

export type InferComponents<A extends Entity[]> = { [K in keyof A]: Type<A[K]> };
```

Reply:
This has been implemented. I appreciate the suggestion.